### PR TITLE
OHRM5X-1127: Timezones API

### PR DESCRIPTION
**Pull request by SandeepaOHRM**
This PR Contains:-

- API for Time zones Autocomplete
- New rule added to validate timezone offset
- Modified Attendance Record API to update employees' time zones
- Modified getTimezoneByOffset helper to get only float as params

### CODE REVIEW

File: symfony/plugins/orangehrmAttendancePlugin/Api/TimezonesAPI.php [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                    }
+                    return false;
+                }
+            );
+        }
+
+        $timezones = [];
+        foreach ($filteredIdentifiers as $timezoneIdentifier) {
+            $timezone = new DateTimeZone($timezoneIdentifier);
+            $offsetInSeconds = $timezone->getOffset(new DateTime());
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAttendancePlugin/Api/TimezonesAPI.php:73]: please check this further

File: symfony/plugins/orangehrmAttendancePlugin/test/fixtures/testcases/TimezonesAPITestCases.yaml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    query:
+      timezoneName: 'tex'
+    data: [ ]
+    meta:
+      total: 0
+    rels: [ ]
+
+
+
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAttendancePlugin/test/fixtures/testcases/TimezonesAPITestCases.yaml:82]: ?

File: symfony/plugins/orangehrmCorePlugin/test/Service/DateTimeHelperServiceTest.php [hunk 3]
```diff
@@ -216,4 +216,11 @@ public function testGetNow()
         $this->assertEquals((new DateTime())->getTimezone(), $this->dateTimeHelperService->getNow()->getTimezone());
         $this->assertEquals((new DateTime())->getOffset(), $this->dateTimeHelperService->getNow()->getOffset());
     }
+
+    public function testGetTimezoneByTimezoneOffset(): void
+    {
+        $this->assertEquals($this->dateTimeHelperService->getTimezoneByTimezoneOffset(5.5), new DateTimeZone('+0530'));
+        $this->assertEquals($this->dateTimeHelperService->getTimezoneByTimezoneOffset(-5.5), new DateTimeZone('-0530'));
+        $this->assertEquals($this->dateTimeHelperService->getTimezoneByTimezoneOffset(5.0), new DateTimeZone('+0500'));
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCorePlugin/test/Service/DateTimeHelperServiceTest.php:224]: `$this->assertEquals($expected, $actual)` @vageeshorangehrm am I wrong?
```suggestion
        $this->assertEquals(new DateTimeZone('+0500'), $this->dateTimeHelperService->getTimezoneByTimezoneOffset(5.0));
```

File: symfony/plugins/orangehrmCorePlugin/test/Service/DateTimeHelperServiceTest.php [hunk 4]
```diff
@@ -216,4 +216,11 @@ public function testGetNow()
         $this->assertEquals((new DateTime())->getTimezone(), $this->dateTimeHelperService->getNow()->getTimezone());
         $this->assertEquals((new DateTime())->getOffset(), $this->dateTimeHelperService->getNow()->getOffset());
     }
+
+    public function testGetTimezoneByTimezoneOffset(): void
+    {
+        $this->assertEquals($this->dateTimeHelperService->getTimezoneByTimezoneOffset(5.5), new DateTimeZone('+0530'));
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCorePlugin/test/Service/DateTimeHelperServiceTest.php:222]: Add more test cases like this
```php
        $this->assertEquals(new DateTimeZone('+0530'), $this->dateTimeHelperService->getTimezoneByTimezoneOffset("5.5"));
        $this->assertEquals(new DateTimeZone('+1100'), $this->dateTimeHelperService->getTimezoneByTimezoneOffset(11));
        $this->assertEquals(new DateTimeZone('-1200'), $this->dateTimeHelperService->getTimezoneByTimezoneOffset(-12));
        $this->assertEquals(new DateTimeZone('-1200'), $this->dateTimeHelperService->getTimezoneByTimezoneOffset(-12.001));
        $this->assertEquals(new DateTimeZone('-0020'), $this->dateTimeHelperService->getTimezoneByTimezoneOffset(-0.333));
```

**Review by RajithaKumara — APPROVED** [review]

File: symfony/plugins/orangehrmCorePlugin/test/Service/DateTimeHelperServiceTest.php (hunk 4 above)
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmCorePlugin/test/Service/DateTimeHelperServiceTest.php:222]: ok

File: symfony/plugins/orangehrmAttendancePlugin/Api/TimezonesAPI.php (hunk 1 above)
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmAttendancePlugin/Api/TimezonesAPI.php:73]: sure

File: symfony/plugins/orangehrmAttendancePlugin/test/fixtures/testcases/TimezonesAPITestCases.yaml (hunk 2 above)
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmAttendancePlugin/test/fixtures/testcases/TimezonesAPITestCases.yaml:82]: Oh sorry!

File: symfony/plugins/orangehrmCorePlugin/test/Service/DateTimeHelperServiceTest.php (hunk 3 above)
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmCorePlugin/test/Service/DateTimeHelperServiceTest.php:224]: public static function assertEquals($expected, $actual, string $message = ''): void
    {
        $constraint = new IsEqual($expected);

        static::assertThat($actual, $constraint, $message);
    }

Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmCorePlugin/test/Service/DateTimeHelperServiceTest.php:224]: sorry I have changed the direction

Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmCorePlugin/test/Service/DateTimeHelperServiceTest.php]: no rajitha you are correct, first argument is expected and second one is actual out put

**Pull request merged**